psydac/core/bsplines.py

```diff
@@ -108,22 +143,51 @@ def basis_funs( knots, degree, x, span ):
     by using 'left' and 'right' temporary arrays that are one element shorter.
 
     """
-    left   = np.empty( degree  , dtype=float )
-    right  = np.empty( degree  , dtype=float )
-    values = np.empty( degree+1, dtype=float )
+    left = np.empty(degree, dtype=float)
+    right = np.empty(degree, dtype=float)
+
+    out[0] = 1.0
+    for j in range(0, degree):
+        left[j] = x - knots[span - j]
+        right[j] = knots[span + 1 + j] - x
+        saved = 0.0
+        for r in range(0, j + 1):
+            temp = out[r] / (right[r] + left[j - r])
+            out[r] = saved + right[r] * temp
+            saved = left[j - r] * temp
+        out[j + 1] = saved
+
+
+# ==============================================================================
+def basis_funs_array(knots: 'float[:]', degree: int, x: 'float[:]', span: 'int[:]', out: 'float[:,:]'):
+    """
+    Compute the non-vanishing B-splines at locations in x, given the knot sequence,
+    polynomial degree and knot span. See Algorithm A2.2 in [1].
 
-    values[0] = 1.0
-    for j in range(0,degree):
-        left [j] = x - knots[span-j]
-        right[j] = knots[span+1+j] - x
-        saved    = 0.0
-        for r in range(0,j+1):
-            temp      = values[r] / (right[r] + left[j-r])
-            values[r] = saved + right[r] * temp
-            saved     = left[j-r] * temp
-        values[j+1] = saved
+    Parameters
+    ----------
+    knots : array_like of floats
+        Knots sequence.
+
+    degree : int
+        Polynomial degree of B-splines.
+
+    x : array_like of floats
+        Evaluation points.
+
+    span : array_like of int
+        Knot span indexes.
+
+    out : array_like of floats
+        Values of p+1 non-vanishing B-Splines at all locations in x.
+
+    """
 
-    return values
+    n = x.shape[0]
+    temp = np.empty(degree + 1, dtype=float)
+    for i in range(n):
+        basis_funs(knots, degree, x[i], span[i], temp)
+        out[i, :] = temp
```

Comment:
There is no need for a temporary array:
```suggestion
    for i in range(n):
        basis_funs(knots, degree, x[i], span[i], out[i, :])
```

Reply:
Fixed in https://github.com/pyccel/psydac/commit/c9eac22256b58bb6b2138192ad3cc5766ae2d143